[Ng Lingshan] iP

lingshanng opened the pull request:
# Duke Task Assistant

> “You can do anything, but not everything.” – David Allen

Duke helps you keep track of all the things you need to do :memo: . It's
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. add your tasks.
3. check your list of tasks anytime
4. you're done with a task, mark it as done

And it's **FREE FOR ALL**  

Features
- [X] Managing todos, events, and deadlines.
- [ ] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

lingshanng committed:
Level 1 Greet Echo Exit

lingshanng committed:
Level 2. Add, List

lingshanng committed:
Level 3. Mark as Done

lingshanng committed:
Level 4. ToDos, Events, Deadlines

lingshanng committed:
add Text UI Tests

lingshanng committed:
add DukeException and handle errors for task details

lingshanng committed:
handle error for incorrect commands

lingshanng committed:
handle task number validation

lingshanng committed:
tweak validateTaskNumber and add deleteTask

lingshanng committed:
fix failed test

lingshanng committed:
add TaskType enum

lingshanng committed:
add documentation

lingshanng committed:
fix trailing spaces for Expected output

lingshanng committed:
make Task abstract and add toDataString method to subclasses

lingshanng committed:
add database methods for init, read and write

lingshanng committed:
store and display deadline date and time as LocalDateTime

lingshanng committed:
update tests for deadline task

lingshanng committed:
Merge branch 'branch-Level-7'

lingshanng committed:
Merge branch 'branch-Level-8'

lingshanng committed:
fix errors for deadline task

lingshanng committed:
refactor - extract Storage

lingshanng committed:
Refactor - extract TaskList

lingshanng committed:
make deadline time AM and PM lowercase

lingshanng committed:
refactor ui displays (wip)

lingshanng committed:
update and fix UI tests

lingshanng committed:
Refactor Parser and Commands, move ui display from TaskList to Commands

lingshanng committed:
add concat method to DukeException

lingshanng committed:
add documentation

lingshanng committed:
divide classes into packages

lingshanng committed:
add JUnit tests

lingshanng committed:
update gitignore for META-INF

lingshanng committed:
add find task function & add tests

lingshanng committed:
capitalise final constants in Deadline

lingshanng committed:
reformat code to adhere to coding standard

lingshanng committed:
Add JavaDocs

lingshanng committed:
Merge branch 'branch-A-CodingStandard'

lingshanng committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/DukeException.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/TaskList.java

lingshanng committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/task/TaskList.java
#	src/test/java/duke/task/TaskListTest.java

lingshanng committed:
use Ui for displayList

zzybluebell commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,35 @@
+
 public class Duke {
     public static void main(String[] args) {
+        String separator = "------------------------------------------------------------------";
```
Perhaps it should use final the string when using over 2 times?

lingshanng replied:
@zzybluebell  Hi I think you are looking at changes from only 1 commit. Maybe try to view all commit changes first and then review from there?

zzybluebell commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,35 @@
+            }
+
+            System.out.println(separator);
+        }
```
Perhaps it can close the scanner :) (But never mind this )

lingshanng committed:
Merge branch 'add-gradle-support' of https://github.com/lingshanng/ip

lingshanng committed:
ise checkStyle to detect coding style violations

lingshanng committed:
set up javaFx and create gui

lingshanng committed:
implement gui with fxml

lingshanng committed:
refactor code to use GUI

lingshanng committed:
Merge branch 'branch-Level-10'

luffingluffy commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    /**
+     * Concats this exception message with another exception message.
```
I think you should spell it out fully.

luffingluffy reviewed (commented):
Well-written and well-documented. LGTM!

CheongYeeMing commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+     * Returns Duke's response to user input.
+     * @param input User's input.
+     * @return Duke's response.
+     */
```
It seems like you have missed an empty line between the description and the `@param`. I might be making an assumption since the rest of your comments are formatted with a space in between the description and `@param` or `@return`.

CheongYeeMing commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,32 @@
+package duke;
+
+public class DukeException extends Exception {
+    /**
```
Missing comments for your  `DukeException` class.

CheongYeeMing commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,77 @@
+        String cmdWord = cmdWordDetails[0]; // get first word as command
+        String description = cmdWordDetails.length > 1 ? cmdWordDetails[1] : "";
+
+        switch (cmdWord) {
+        case "bye":
+            return new ExitCommand();
+        case "list":
+            return new DisplayCommand();
+        case "todo":
+            return new AddCommand(TaskList.TaskType.TODO, description);
+        case "deadline":
+            return new AddCommand(TaskList.TaskType.DEADLINE, description);
+        case "event":
+            return new AddCommand(TaskList.TaskType.EVENT, description);
+        case "done":
+            return new DoneCommand(validateTaskNumber(description, tasks));
+        case "delete":
+            return new DeleteCommand(validateTaskNumber(description, tasks));
+        case "find":
+            return new FindCommand(description);
+        default:
+            throw new DukeException("Sorry, I don't know what that means.");
```
Different commands are neatly dealt with! 👍

CheongYeeMing commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,40 @@
+     * @return true if command is exit command.
+     */
+    public abstract boolean isExit();
+    
+    /** Append multiple strings into one string separated by newline character. */
+    protected String formatOutput(String... outputs) {
+        return String.join("\n", outputs);
+    }
+
+    /** Append multiple strings in an array into one string separated by newline character. */
+    protected String formatOutput(ArrayList<String> outputs) {
+        return String.join("\n", outputs);
+    }
+}
```
Comments for these 2 methods missing the `@param` and `@return` sections.

CheongYeeMing commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,49 @@
+     * @return a string output.
+     * @throws DukeException
+     */
+    public String execute(TaskList tasks, Storage storage) throws DukeException {
+        Task task = tasks.markTaskDone(taskNum);
+        
+        storage.save(tasks.getListData());
+        return formatOutput("Good work! I've marked this task as done:", task.toString());
+    }
```
Minor formatting discrepancy. Compare with the `execute` method of `DeleteCommand`.

CheongYeeMing reviewed (commented):
Overall very clean code. Just some minor discrepancies in comments and formatting of similar methods, for example, the `execute` method in `DeleteCommand` and `DoneCommand`. Good job!!

zzybluebell commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -10,15 +13,25 @@ public ToDo(String description) throws DukeException {
         return String.join(delimiter, tag, done, super.description);
     }
-}
+}
```
Overall Looks very good, awesome

zzybluebell commented on src/test/java/duke/task/ToDoTest.java:
```diff
@@ -0,0 +1,41 @@
+        } catch (DukeException e) {
+            fail();
+        }
+    }
```
Very sufficient Junit test case

lingshanng committed:
update UI

lingshanng committed:
update build.gradle

lingshanng committed:
Add assertions in Storage.load() and TaskList methods.

* For TaskList methods markTaskDone() and deleteTask(), todoList
should have at least 1 task to work correctly.
* For Storage.load(), the task data parsed into details is
assumed to be formatted correctly containing 3 fields for ToDo
and 4 fields for Deadline and Event.

The assumptions may cause bugs and errors if not met.

Adding assertions will allow assumptions to be verified in runtime
and surface bugs if any.

Using assertions in TaskList is preferable over Exceptions as
the failed validation of taskNum should have already thrown
an Exception in Command before calling the methods in TaskList.

lingshanng committed:
Extract and consolidate duplicate code for conditional fragments.

lingshanng committed:
fix incorrect error message

lingshanng committed:
use stream to filter todoList for tasks matching the key

lingshanng committed:
Create gradle.yml

lingshanng committed:
fix checkstyle errors and tests

lingshanng committed:
Merge branch 'master' into branch-A-Assertions

lingshanng committed:
Merge branch 'master' into branch-A-CodeQuality

lingshanng committed:
Merge branch 'master' into branch-A-Streams

lingshanng committed:
fix checkstyle error with line length

lingshanng committed:
fix checkstyle error with trailing spaces

lingshanng committed:
Merge pull request #3 from lingshanng/branch-A-Streams

Use Streams

lingshanng committed:
Merge branch 'master' into branch-A-Assertions

lingshanng committed:
Merge pull request #1 from lingshanng/branch-A-Assertions

Add assertions

lingshanng committed:
Merge branch 'master' into branch-A-CodeQuality

lingshanng committed:
Merge pull request #2 from lingshanng/branch-A-CodeQuality

Refactor duplicate code

lingshanng committed:
tweak ui and add a help window

lingshanng committed:
add a help command to show list of available commands

lingshanng committed:
Merge branch 'branch-C-Help'

lingshanng committed:
hide windows on exit

lingshanng committed:
improve UI and change welcome message

lingshanng committed:
Use Textflow and Text instead of Label, add copy text feature

lingshanng committed:
Merge branch 'branch-A-BetterGui'

lingshanng committed:
Set theme jekyll-theme-cayman

lingshanng committed:
add monospace font for code block, use code block for command list

lingshanng committed:
add Ui.png

lingshanng committed:
Merge branch 'master' of https://github.com/lingshanng/ip

lingshanng committed:
fix javadoc for isShowing method

lingshanng committed:
update Ui.png

lingshanng committed:
Update Features

lingshanng committed:
Update User Guide

lingshanng committed:
Update User Guide

lingshanng committed:
Update note

lingshanng committed:
Update note

lingshanng committed:
Add content list

lingshanng committed:
Update README.md

lingshanng committed:
Update README.md

lingshanng committed:
Merge branch 'master' of https://github.com/lingshanng/ip

lingshanng committed:
update help documentation with find

lingshanng committed:
Update README.md

lingshanng committed:
Update README.md

lingshanng committed:
move app_data to resources and rename jar file

lingshanng committed:
Merge branch 'master' of https://github.com/lingshanng/ip
